Create "cache" folder if it doesn't exist before trying to write to it

AndySDH opened the pull request:
This fixes this issue:

### Steps to reproduce the issue

- Have cache enabled
- Delete cache folder

### Result before PR

Sites breaks. Goes blank.

With error reporting on, throws error:
"Error: Call to a member function getTag() on null: The file Cache Storage is not supported on this platform."

### Result after PR

Joomla creates "cache" folder if it doesn't exist

### Additional comments

Now, of course there is no reason why one should delete the cache folder but it can happen by accident or for whatever other reason. The principle that Joomla doesn't check if the folder is there or not (and breaks the site if it's not) before trying to write to it is wrong.

AndySDH committed:
Create "cache" folder if it doesn't exist before trying to write to it 

This fixes this issue:

### Steps to reproduce the issue

- Have cache enabled
- Delete cache folder

### Expected result

- Joomla creates "cache" folder if it doesn't exist

### Actual result

- Sites breaks. Goes blank.

### Additional comments

Now, of course there is no reason why one should delete the cache folder but it can happen by accident or for whatever other reason. The principle that Joomla doesn't check if the folder is there or not (and breaks the site if it's not) before trying to write to it is wrong.

brianteeman commented on libraries/src/Cache/Storage/FileStorage.php:
```diff
@@ -334,7 +334,14 @@ public function gc()
+		if (!is_dir($folder))
+		{
+			@mkdir($folder);
+		}
```
Don't we need some code then if for any reason joomla cannot create the folder

mbabker commented:
This is the wrong place to be doing this type of thing.  A check for platform support shouldn't have side effects, in this case the side effect is creating the directory.

This type of directory creation is better suited in com_config when the global configuration is saved (which, we already do except for trying to (re-)create the core cache directory).

TBH, the more I think about this the more I think this particular `isSupported` method should just return true and leave it to runtime to raise issues if there are problems reading from or writing to the configured cache directory (we don't try to create the root cache path at runtime, which IMO should correctly be an error, but we do create the subdirectories as needed).

mbabker commented:
Taking some inspiration from https://github.com/joomla-framework/cache/blob/2.0-dev/src/Adapter/File.php

1) Make the `isSupported` method only return true, don't check config or anything like that (especially because the cache path is injectable in the handler's constructor, which may be different from the config path)
2) Check that the path exists and is readable in the handler's constructor, if not throw a `CacheConnectingException` (this is in line with how we fail on the Memcached handler if the connection can't be created at runtime)
3) The same changes should be made to the Cache_Lite handler (it too is a filesystem cache API just using a third party PEAR library)

AndySDH committed:
Update FileStorage.php

By removing the isSupported method from that file, it will fall back on its parent isSupported method, which simply returns true.

AndySDH commented:
@mbabker Thanks for the comment. Deleting the method solves the issue too. Updated.
Beyond this, I'm not that knowledgeable, so this is as far as I can go with this. If you want to implement a different/more thorough solution, please go ahead.
But let's not take weeks for a simple fix, this is already better than before when it was breaking the site.

franz-wohlkoenig commented:
@HLeithner do you plan to merge this in J3 or is it for J4?

HLeithner commented:
I will merge it. @AndySDH please resolve the conflict

AndySDH committed:
Merge branch 'staging' into patch-1

AndySDH commented:
Done (I think. First time I resolve conflicts stuff)

AndySDH committed:
Update FileStorage.php

HLeithner commented:
thx
